mapiproxy/servers/default/emsmdb/dcesrv_exchange_emsmdb.c

```diff
@@ -1906,9 +1536,58 @@ static enum MAPISTATUS dcesrv_EcDoAsyncConnectEx(struct dcesrv_call_state *dce_c
 						 TALLOC_CTX *mem_ctx,
 						 struct EcDoAsyncConnectEx *r)
 {
-	OC_DEBUG(3, "exchange_emsmdb: EcDoAsyncConnectEx (0xe) not implemented\n");
-	r->out.result = ecRejected;
-	DCESRV_FAULT(DCERPC_FAULT_OP_RNG_ERROR);
+	struct exchange_emsmdb_session	*session;
+	enum mapistore_error		retval;
+	struct emsmdbp_context		*emsmdbp_ctx;
+	struct dcesrv_handle		*handle;
+
+	OC_DEBUG(3, "exchange_emsmdb: EcDoAsyncConnectEx (0xe)\n");
+
+	/* Step 0. Ensure incoming user is authenticated */
+	if (!dcesrv_call_authenticated(dce_call)) {
+		OC_DEBUG(1, "No challenge requested by client, cannot authenticate");
+		r->out.async_handle->handle_type = 0;
+		r->out.async_handle->uuid = GUID_zero();
+		r->out.result = DCERPC_FAULT_CONTEXT_MISMATCH;
+		return MAPI_E_LOGON_FAILED;
+	}
+
+	/* Step 1. Retrieve the existing session */
+	session = dcesrv_find_emsmdb_session(&r->in.handle->uuid);
+	if (session) {
+		emsmdbp_ctx = (struct emsmdbp_context *) session->session->private_data;
+	} else {
+		OC_DEBUG(0, "[EcDoAsyncConnectEx]: emsmdb session not found");
+		r->out.async_handle->handle_type = 0;
+		r->out.async_handle->uuid = GUID_zero();
+		r->out.result = DCERPC_FAULT_CONTEXT_MISMATCH;
+		return MAPI_E_LOGON_FAILED;
+	}
+
+	/* Step 2. Create the new session for async pipe */
+	handle = dcesrv_handle_new(dce_call->context, EXCHANGE_HANDLE_ASYNCEMSMDB);
+	if (!handle) {
+		OC_DEBUG(0, "[EcDoAsyncConnectEx] Unable to create new handle");
+		r->out.async_handle->handle_type = 0;
+		r->out.async_handle->uuid = GUID_zero();
+		r->out.result = DCERPC_FAULT_CONTEXT_MISMATCH;
+		return MAPI_E_LOGON_FAILED;
+	}
+
+	/* Step 3. Register the global session */
+	retval = mapistore_notification_session_add(emsmdbp_ctx->mstore_ctx, r->in.handle->uuid,
```

Comment:
there is no difference, but for me it would be more clear if we use `emsmdbp_ctx->session_uuid` instead of `r->in.handle->uuid`.
And instead of `dcesrv_call_account_name(dce_call)` `emsmdbp_ctx->username`